File: forest_lib/forest/image.py
"""

Image comparison tools
======================

Tools to compare RGBA images on a pixel by pixel basis

References to the :class:`bokeh.models.sources.ColumnDataSource` sources
related to the left/right images should be provided by the user. This
is typically the responsibility of some glue code, either a ``main.py``
or some other calling code

.. note:: The algorithms and data structures expose methods
          that can be attached to bokeh events. They are
          unaware of the calling code architecture and simply
          edit appropriate RGBA values

Slider
------

A :class:`.Slider` is available that makes it easy to compare two
images relative to a mouse pointer. The slider splits images
such that the left portion of one image and
the right portion of the other image are visible either side of the mouse
position.

>>> slider = forest.image.Slider(left_images, right_images)
>>> slider.add_figure(figure)

Application programming interface (API)
---------------------------------------

The following classes have been made available to users
of Forest for custom visualisations
"""
import os
import sys
import numpy as np
import bokeh.models
import bokeh.events


# CustomJS callback code
JS_FILE = os.path.join(os.path.dirname(os.path.realpath(__file__)),
                       "image.js")
with open(JS_FILE, "r") as stream:
    JS_CODE = stream.read()


class Slider(object):
    """Controls alpha values of bokeh RGBA ColumnDataSources

    Understands bokeh architecture, either a GlyphRenderer
    or a ColumnDataSource may be passed in for either
    left_images or right_images arguments

    The various properties needed to set alpha values relative
    to the mouse position can be found by inspecting
    either data structure

    .. note:: If a ColumnDataSource is passed then the following
              keys are used to find the image properties
              'image', 'x', 'y', 'dw', 'dh'

    :param left_images: ColumnDataSource or GlyphRenderer used to
                        define RGBA images to the left of the cursor
    :param right_images: ColumnDataSource or GlyphRenderer used to
                         define RGBA images to the right of the cursor
    """

    def __init__(self, left_images, right_images):
        self.images = {
            "left": left_images,
            "right": right_images
        }
        self.span = bokeh.models.Span(location=0,
                                      dimension='height',
                                      line_color='black',
                                      line_width=1)
        self.shared = bokeh.models.ColumnDataSource({
            "use_previous_mouse_x": [False],
            "previous_mouse_x": [0]
        })
        self.mousemove = bokeh.models.CustomJS(args=dict(
            span=self.span,
            left_images=self.left_images,
            right_images=self.right_images,
            shared=self.shared
        ), code=JS_CODE)

    @property
    def left_images(self):
        return self.images["left"]

    @property
    def right_images(self):
        return self.images["right"]

    def add_figure(self, figure):
        """Attach various callbacks to a particular figure"""
        self.hover_tool = bokeh.models.HoverTool(callback=self.mousemove)
        figure.add_tools(self.hover_tool)
        figure.renderers.append(self.span)

        # Add MouseEnter callback to reset slider
        callback = bokeh.models.CustomJS(args=dict(shared=self.shared), code="""
            shared.data.use_previous_mouse_x[0] = false;
        """)
        figure.js_on_event(bokeh.events.MouseEnter, callback)


def get_shapes(source):
    """Read image shapes from ColumnDataSource"""
    images = source.data["image"]
    return [image.shape for image in images]


def get_alpha(bokeh_obj):
    """Helper to copy alpha values from GlyphRenderer or ColumnDataSource"""
    images = get_images(bokeh_obj)
    return [np.copy(rgba[..., -1]) for rgba in images]


def get_images(bokeh_obj):
    """Helper to get image data from ColumnDataSource or GlyphRenderer"""
    if hasattr(bokeh_obj, "data_source"):
        renderer = bokeh_obj
        return renderer.data_source.data[renderer.glyph.image]
    column_data_source = bokeh_obj
    return column_data_source.data["image"]


def alpha_from_rgba(rgba):
    """Helper method to view alpha values from an RGBA array

    The alpha values are assumed to be the last entry of the
    last dimension

    .. note:: This method does not handle RGBA data stored as
              1D arrays

    :param rgba: Array of RGBA values
    :returns: View on alpha values
    """
    return rgba[..., -1]

File: forest_lib/forest/image.js
"use strict";
// bokeh.models.CustomJS callback code
//
// args:
//      span - bokeh.models.Span
//      left_images - ColumnDataSource
//      left_alpha - Copy of left_images alpha values
//      right_images - ColumnDataSource
//      right_alpha - Copy of right_images alpha values
//      shared - ColumnDataSource
// cb_data:
//      geometry.x - mouse x position relative to figure
//
let reveal_image = function(side,
                            source,
                            mouse_x,
                            previous_mouse_x,
                            use_previous_mouse_x) {
    // RGBA image extents in mouse position space
    let x = source.data["x"][0];
    let y = source.data["y"][0];
    let dw = source.data["dw"][0];
    let dh = source.data["dh"][0];
    let original_alpha = source.data["_alpha"][0];
    let ni = source.data["_shape"][0][0];
    let nj = source.data["_shape"][0][1];

    // Useful debug information
    let mode = "silent";
    if (mode === "debug") {
        console.log(side, "x", x);
        console.log(side, "y", y);
        console.log(side, "dw", dw);
        console.log(side, "dh", dh);
        console.log(side, "alpha", original_alpha[0]);
        console.log(side, "ni", ni);
        console.log(side, "nj", nj);
    }

    // Mouse position(s)
    let left_x;
    let right_x;
    if (!isFinite(mouse_x)) {
        return;
    }
    if (mouse_x > previous_mouse_x) {
        left_x = previous_mouse_x;
        right_x = mouse_x;
    } else {
        left_x = mouse_x;
        right_x = previous_mouse_x;
    }

    // Update alpha pseudo-2D and RGBA pseudo-3D arrays
    let pixel_x;
    let alpha;
    let alpha_index;
    let image_alpha_index;
    let original_alpha_value;
    let dy = dw / nj;
    let skip = 0;

    for (let j=0; j<nj; j++) {
        pixel_x = x + (j * dy);

        // Optimised selection of columns between mouse events
        // note: feature turned off during first paint
        if (use_previous_mouse_x) {
            if ((pixel_x > right_x) || (pixel_x < left_x)) {
                // pixel outside current and previous mouse positions
                skip += 1;
                continue;
            }
        }

        // Ordinary loop logic
        for (let i=0; i<ni; i++) {
            alpha_index = (nj * i) + j;
            original_alpha_value = original_alpha[alpha_index];
            if (original_alpha_value == 0) {
                continue;
            }
            image_alpha_index = (4 * alpha_index) + 3;
            if (visible_pixel(pixel_x, mouse_x, side)) {
               alpha = original_alpha_value;
            } else {
               alpha = 0;
            }
           source.data["image"][0][image_alpha_index] = alpha;
        }
    }
    if (skip !== nj) {
        // Some columns need to be painted
        source.change.emit();
    }
};

let visible_pixel = function(pixel_x, mouse_x, show_side) {
    if (show_side === "left") {
        return pixel_x < mouse_x;
    } else {
        return pixel_x > mouse_x;
    }
};

// CustomJS callback main program
let main = function(cb_data,
                    left_images,
                    right_images,
                    shared,
                    span) {
    // Gather data from cb_data and args
    let mouse_x = cb_data.geometry.x;
    let previous_mouse_x = shared.data.previous_mouse_x[0];
    let use_previous_mouse_x = shared.data.use_previous_mouse_x[0];

    // Move vertical line to mouse position
    span.location = mouse_x;

    // Update image alpha values
    reveal_image("left",
                 left_images,
                 mouse_x,
                 previous_mouse_x,
                 use_previous_mouse_x);
    reveal_image("right",
                 right_images,
                 mouse_x,
                 previous_mouse_x,
                 use_previous_mouse_x);

    // Update shared data
    shared.data.previous_mouse_x[0] = mouse_x;
    shared.data.use_previous_mouse_x[0] = true;
};

if (typeof module === 'undefined') {
    // Bokeh call back usage
    main(cb_data,
         left_images,
         right_images,
         shared,
         span);
} else {
    // NPM test usage
    module.exports = {
        main: main,
        reveal_image: reveal_image,
        visible_pixel: visible_pixel
    };
}
